core/trace/src/test/scala/org/typelevel/otel4s/trace/BaseSpanBuilderSuite.scala

```diff
@@ -0,0 +1,268 @@
+    b = b.withParent(spanContext)
+    b = b.withFinalizationStrategy(finalizationStrategy)
+    b = b.withSpanKind(spanKind)
+    b = b.withStartTimestamp(timestamp)
+
+    val ops = b.build
+
+    for {
+      _ <- IO(assert(!builder.meta.isEnabled))
+      _ <- ops.use_
+    } yield assert(!allocated)
+  }
+
+  // if the optimization is not applied, the compilation will fail with
+  // Method too large: org/typelevel/otel4s/...
+  // Scala 3 compiler optimizes chained calls out of the box
+  // But Scala 2 doesn't, see SpanBuilderMacro#whenEnabled for manual optimization logic
+  builderTest("optimize chained calls", enabled = false) { builder =>
+    var allocated = false
+
+    def timestamp: FiniteDuration = {
+      allocated = true
+      100.millis
+    }
+
+    def attribute: Attribute[String] = {
+      allocated = true
+      Attribute("key", "value")
+    }
+
+    def attributes: List[Attribute[String]] = {
+      allocated = true
+      List(Attribute("key", "value"))
+    }
+
+    def spanContext: SpanContext = {
+      allocated = true
+      SpanContext.invalid
+    }
+
+    def finalizationStrategy: SpanFinalizer.Strategy = {
+      allocated = true
+      SpanFinalizer.Strategy.empty
+    }
+
+    def spanKind: SpanKind = {
+      allocated = true
+      SpanKind.Client
+    }
+
+    val ops = builder
+      .addAttribute(attribute)
+      .addAttributes(attribute, attribute)
+      .addAttributes(attributes)
+      .addLink(spanContext, attribute, attribute)
+      .addLink(spanContext, attributes)
+      .withParent(spanContext)
+      .withFinalizationStrategy(finalizationStrategy)
+      .withSpanKind(spanKind)
+      .withStartTimestamp(timestamp)
+      .addAttribute(attribute)
+      .addAttributes(attribute, attribute)
+      .addAttributes(attributes)
+      .addLink(spanContext, attribute, attribute)
+      .addLink(spanContext, attributes)
+      .withParent(spanContext)
+      .withFinalizationStrategy(finalizationStrategy)
+      .withSpanKind(spanKind)
+      .withStartTimestamp(timestamp)
+      .addAttribute(attribute)
+      .addAttributes(attribute, attribute)
+      .addAttributes(attributes)
+      .addLink(spanContext, attribute, attribute)
+      .addLink(spanContext, attributes)
+      .withParent(spanContext)
+      .withFinalizationStrategy(finalizationStrategy)
+      .withSpanKind(spanKind)
+      .withStartTimestamp(timestamp)
+      .addAttribute(attribute)
+      .addAttributes(attribute, attribute)
+      .addAttributes(attributes)
+      .addLink(spanContext, attribute, attribute)
+      .addLink(spanContext, attributes)
+      .withParent(spanContext)
+      .withFinalizationStrategy(finalizationStrategy)
+      .withSpanKind(spanKind)
+      .withStartTimestamp(timestamp)
+      .build
+
+    for {
+      _ <- IO(assert(!builder.meta.isEnabled))
+      _ <- ops.use_
+    } yield assert(!allocated)
+  }
+
+  builderTest("should be stack-safe") { builder =>
```

Comment:
the failure mode of this test is awful (putting the VM into an unusable/broken state) but I don't know that there's a better way to do it. we should probably be forking the VM for tests if we weren't already though